pandas/core/arrays/datetimelike.py

```diff
@@ -1703,6 +1704,13 @@ def _groupby_op(
         res_values = res_values.view(self._ndarray.dtype)
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other_for_merge(
+        self, other: Self, sort: bool = False
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        lk = np.asarray(self._ndarray, dtype=np.int64)
```

Comment:
`# caller is responsible for ensuring self.dtype == other.dtype`

Reply:
added